### PR TITLE
feat(manager/azure-pipelines): extract azure-pipelines tasks

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes

Adds support for extracting Azure Pipelines tasks.

Azure Pipelines YAML has two optional levels of hierarchy, `stages`, and `jobs`, and one required level, `steps`. These changes allow Renovate to extract `task` configurations from any of the levels.

## Context

- First step towards #15818 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/azure-pipelines/__fixtures__/azure-pipelines-alias.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+steps:
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/__fixtures__/azure-pipelines-alias.yaml:1]: please inline these small fixtures

File: lib/modules/manager/azure-pipelines/__fixtures__/azure-pipelines-invalid.yaml [hunk 2]
```diff
@@ -1,3 +1 @@
-steps:
-- bash: echo "hello"
-  displayName: echo
+}
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/__fixtures__/azure-pipelines-invalid.yaml:1]: ☝️

File: lib/modules/manager/azure-pipelines/extract.spec.ts [hunk 3]
```diff
@@ -88,21 +94,69 @@ describe('modules/manager/azure-pipelines/extract', () => {
     });
   });
 
+  describe('extractAzurePipelinesTasks()', () => {
+    it('should extract azure-pipelines task information', () => {
+      expect(extractAzurePipelinesTasks('Bash@3')).toMatchSnapshot({
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.spec.ts:99]: no new snapshots

https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 4]
```diff
@@ -83,18 +93,64 @@ export function extractPackageFile(
...
+      const dep = extractContainer(container);
+      if (dep) {
+        deps.push(dep);
+      }
+    }
+  }
+
+  if (!is.nullOrUndefined(pkg.stages)) {
+    const tasks = pkg.stages
+      .flatMap((stage) =>
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.ts:123]: please prefer `for...of` loops

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/azure-pipelines/extract.spec.ts (hunk 3 above)
Comment (JamieMagee) [reply lib/modules/manager/azure-pipelines/extract.spec.ts:99]: I got rid of the existing snapshots as well 🎉

File: lib/modules/manager/azure-pipelines/extract.ts (hunk 4 above)
Comment (JamieMagee) [reply lib/modules/manager/azure-pipelines/extract.ts]: I think it gets quite deeply nested, but I've made the change.

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 5]
```diff
@@ -83,18 +93,67 @@ export function extractPackageFile(
...
+    !is.nullOrUndefined(pkg.resources) &&
+    is.nonEmptyArray(pkg.resources.containers)
+  ) {
+    for (const container of pkg.resources.containers) {
+      const dep = extractContainer(container);
+      if (dep) {
+        deps.push(dep);
+      }
+    }
+  }
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.ts:119]:
```suggestion 
  for (const container of pkg.resources?.containers ?? []) {
    const dep = extractContainer(container);
    if (dep) {
      deps.push(dep);
    }
  }
```

reduce nesting and simplify code?

**Review by viceice — COMMENTED** [review]
I would recommend to do same for the other cases to reduce code nesting.

**Review by viceice — APPROVED** [review]

**Pull request merged**